### PR TITLE
Make cluster_dump available for other clusters

### LINKED ISSUE #184: Make cluster_dumps available beyond the scope of small_client fixture
See discussion https://github.com/coiled/coiled-runtime/issues/177#issuecomment-1156716714

### PR DESCRIPTION
**Pull request by ncclementi**
- [x] Closes #184

### CODE REVIEW

**Comment by ncclementi** [comment]
@ian-r-rose it looks like something different is happening when launching a cluster via coiled, that the plugin doesn't work as it worked locally. 

At this point I'm not sure what's going on, I would have expect "failed" to be true in this case and not try to compute `request.node.rep_call.failed` but it didin't. Any extra ideas on what could be happening? 
https://github.com/coiled/coiled-runtime/runs/6993735017?check_suite_focus=true#step:6:150

File: conftest.py [hunk 1]
```diff
@@ -170,3 +149,40 @@ def s3_cluster_dump_url(s3, s3_scratch):
...
+        # the code below is a workaround to make cluster dumps work with clients in fixtures
+        # and outside fixtures.
+        try:
+            yield
+        except:  # noqa: E722
+            failed = True
+            raise
+        finally:
+            cluster_dump = strtobool(os.environ.get("CLUSTER_DUMP", "false"))
+            logger.error(f"Print failed var for testing purposes, failed=: {failed}")
```
[the remark this answers is not in the document]
Comment (ncclementi) [reply conftest.py:187]: The new logic I added works when the test passes, see https://github.com/coiled/coiled-runtime/runs/7013741536?check_suite_focus=true

I'm testing a failure now and see if this works.

File: conftest.py [hunk 2]
```diff
@@ -170,3 +149,46 @@ def s3_cluster_dump_url(s3, s3_scratch):
...
+        try:
+            yield
+        except:  # noqa: E722
+            failed = True
+            raise
+        finally:
+            try:
+                # we need this for tests that are not using the client fixture
+                # for those cases request.node.rep_call.failed can't be access.
+                failed_req = request.node.rep_call.failed
```
Comment (ian-r-rose) [inline conftest.py:184]: Can this just be
```suggestion
                failed = request.node.rep_call.failed
```
?

Do we need the extra flag?

File: conftest.py [hunk 3]
```diff
@@ -170,3 +149,46 @@ def s3_cluster_dump_url(s3, s3_scratch):
...
+@pytest.fixture
+def upload_cluster_dump(request, s3_cluster_dump_url, s3_storage_options):
+    @contextlib.contextmanager
+    def _upload_cluster_dump(client, cluster):
+        failed = False
+        # the code below is a workaround to make cluster dumps work with clients in fixtures
+        # and outside fixtures.
+        try:
+            yield
+        except:  # noqa: E722
```
Comment (ian-r-rose) [inline conftest.py:177]:
```suggestion
        except Exception:
```

File: conftest.py [hunk 4]
```diff
@@ -170,3 +149,46 @@ def s3_cluster_dump_url(s3, s3_scratch):
...
+            failed = True
+            raise
+        finally:
+            try:
+                # we need this for tests that are not using the client fixture
+                # for those cases request.node.rep_call.failed can't be access.
+                failed_req = request.node.rep_call.failed
+            except AttributeError:
+                failed_req = False
```
Comment (ian-r-rose) [inline conftest.py:186]:
```suggestion
                failed = False
```

**Review by ian-r-rose — COMMENTED** [review]
Thanks @ncclementi, this looks good to me, just some minor suggestions

**Pull request merged**